### ISSUE TITLE
How to run migrations on database created when tenant is created ?

### ISSUE THREAD

**Issue by LemarinelNet** [spatie/laravel-multitenancy#100]
Hi

I'm struggling a bit with this !
I got my database created automatically when a tenant is created, with an Event listener on my Tenant model : 
```
        static::created(function (UserTenant $model) {
            $model->createDatabase();
        });

       // ....
    public function createDatabase()
    {
      DB::connection()->statement("CREATE DATABASE " . $this->getDatabaseName() );

     Artisan::call('tenants:artisan "migrate --database=tenant" --tenant=' . $this->id);
```
and I tried to add some call to the Artisan command from this function, so the table are migrated, but I get an error  : 
```'There are no commands defined in the "migrate" namespace.'```

Thanks for your help !

**Comment by ArtisanTinkerer** [comment]
This is working for me:
```
Artisan::call("tenants:artisan \"migrate --database=tenant\" --tenant={$tenant->id}");
```

**Comment by LemarinelNet** [comment]
Thanks for you reply,

That seems to be the same line than mine, except you used double quotes...
Where are you running this code in ?
I do in my Tenant Model, which is not based on Command class. Perhaps your is located in a command ?

**Comment by ArtisanTinkerer** [comment]
I think it is the quotes, try it like mine. If I use your line in my code, I get the same error.

**Comment by masterix21** [comment]
@LemarinelNet seems to be a parameters problem. Try with:
```php
Artisan::call('tenants:artisan', [
    "artisanCommand" => "migrate --database=tenant",
    "--tenant" => $this->id,
]);
```

**Comment by LemarinelNet** [comment]
Still the same error :
`Symfony/Component/Console/Exception/NamespaceNotFoundException with message 'There are no commands defined in the "migrate" namespace.'`

My full class is : 
```
<?php

namespace App;

use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\Facades\DB;
use Spatie\Multitenancy\Models\Tenant;

class UserTenant extends Tenant
{
    protected $table = "tenants";

    public static function booted()
    {
        static::created(function (UserTenant $model) {
            $model->createDatabase();
        });
    }

    public function createDatabase()
    {
        $this->database = "tenantdb_" . uniqid();
        $this->save();

        $db = DB::connection();

        $charset = $db->getConfig('charset');
        $collation = $db->getConfig('collation');

        $db->statement('CREATE DATABASE ' . $this->getDatabaseName() . " CHARACTER SET `$charset` COLLATE `$collation`");

        Artisan::call('tenants:artisan', [
            "artisanCommand" => "migrate --database=tenant",
            "--tenant" => $this->id,
        ]);
    }

}

```

Could it be that the Artisan is searching a "migrate" command in the current namespace App, but it only exists in the App\Console\Commands namespace ?

**Comment by masterix21** [comment]
I don't know why it happens, but can you try to execute the following code before your migrations?

```php
Application::starting(function ($artisan) { 
    $artisan->resolveCommands([ 
        \Illuminate\Database\Console\Migrations\MigrateCommand::class,
    ]);
});
```

**Comment by LemarinelNet** [comment]
Hi,
This command doesn't fail, but doesn't resolve to anything and doesn't return anything ? 
What should it do ?
Placing it before my migration command still doesn't resolve the error.

**Comment by masterix21** [comment]
It says to your application how to resolve your migration command. I had a similar problem fixed using the suggested code. 

I think that your problem isn't package-dependent, sorry.

**Comment by LemarinelNet** [comment]
Finaly solved it !
My initial code was in fact working, but the issue was from Tinker, which doesn't allow migrate commands in the shell (see https://github.com/laravel/tinker/issues/37).
Tinker has a white list of allowed commands to be run, and migrate is not among them !
So, it was working from the application (web user registration), but not from tinker !

I removed the "Application::starting" part that you suggested, since it was making my code failing, and just the Artisan::call did the trick !

Thanks for your help anyway ;)

**Comment by masterix21** [comment]
Great! Thanks for your share

**Comment by schiffty** [comment]
@LemarinelNet Thanks for the tip on the whitelist, that solved my problem too. 

One question, when you added `migrate:install` to the whitelist, did you do it through TinkerCommand.php or through the tinker.php config file? I kept getting a resolution error when trying to do it the "right" way by adding the InstallCommand class to tinker.php, so I fell back to adding `migrate:install` directly to TinkerCommand.php to get it working and just want to confirm I'm not missing something obvious. Thanks.

**Comment by joselara** [comment]
I'm having the same issue using Artisan call method inside the console. It looks like it has something todo with tinker itself.

```Symfony/Component/Console/Exception/NamespaceNotFoundException with message 'There are no commands defined in the "migrate" namespace.'```

```
 Artisan::call('tenants:artisan', [
     "artisanCommand" => 'migrate --path=database/migrations/tenant --database=tenant',
     "--tenant" => $this->id
])
```

**Comment by masterix21** [comment]
@joselara are you calling the migration from tinker or with a console command? Please post your code. 

Thanks

**Comment by michaelhume** [comment]
Hi @schiffty, et. al.. - did you ever resolve your question? I've landed here having the same issue. I've tried to add

```
'commands' => [
        \Illuminate\Database\Console\Migrations\MigrateCommand::class,
        \Spatie\Multitenancy\Commands\TenantsArtisanCommand::class,
    ],
```
to `tinker.php` without any luck. I still get the `'There are no commands defined in the "migrate" namespace.'` error when trying to run code that contains Artisan::call from within tinker.

@masterix21 - my code is inside a job that is being triggered within tinker when creating a new tenant.

**Comment by masterix21** [comment]
Hi @michaelhume, if you can, please share your code in a testable GitHub repository: I'll take a look asap.

**Comment by michaelhume** [comment]
Thanks @masterix21 - I'm digging into it a bit more and will certainly post either a solution or a testable repo.

**Comment by michaelhume** [comment]
I was able to sort it all out. There were a number of small configuration issues but ultimately attempting to use a database queue driver in a multitenant setup and dispatching jobs on new models that don't yet have a database gets a bit hairy. 

Thanks!

**Comment by michaelhume** [comment]
Hi @masterix21 - Just to add a bit to this as it is coming up again for me. I've posted a question over on [laracasts](https://laracasts.com/discuss/channels/laravel/there-are-no-commands-defined-in-the-migrate-namespace?page=1) as well, but wanted to mention here as I _suspect_ this may be related to this issue.

TLDR - I believe `'There are no commands defined in the "migrate" namespace.'` is a red herring and the issue is actually somehow with the [Migrator](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Migrations/Migrator.php) database connection. I can toggle this error by manipulating `app('migrator')->setConnection()` and running `migrate` in tinker. By setting to the tenant connection, I always get the error, and can clear it be setting the landlord connection. This is regardless of having a current tenant, or fixing a tenant DB_DATABASE in config. 

I don't think it's really a package issue because this works when not in tinker, but I'm just wondering if anyone else has run into this or has any insight into what might be happening here? I'd like to understand why the different behaviour inside tinker.

Thanks!
